### DISCUSSION TITLE
Support for AU Plugins

### DISCUSSION THREAD

**Discussion by xrad** [ExistentialAudio/BlackHole#18] [Feature Requests]
It would be very nice to be able to stick AU plugins in the processing chain to support things like [AutoEq](https://github.com/jaakkopasanen/AutoEq). SoundSource is an commercial solution for this and I'm using it, but a free alternative would be great to have!

**Comment by devinroth** [comment]
Thanks. I’ll add it to the suggestions.

**Comment by xrad** [comment]
Awesome! Let me know if you need any input.

**Comment by devinroth** [comment]
Will do. Probably won't happen for a while.

**Comment by leocoracini** [comment]
Please @devinroth It will definitely be a break deal with AU Plugins. The thing is that just a little percentage of people know how amazing AUGraphicEQ is for the hole output audio on MacOS.
Hope it won't be difficult to build 😊
Thanks

**Comment by xrad** [comment]
@leocoracini In the meantime you can use SoundSource. Commercial, but not overly expensive IMO.

**Comment by danekszy** [comment]
Doesn't AU Lab handle such scenarios?

**Comment by devinroth** [comment]
Yeah AU Lab will do the trick.

**Comment by devinroth** [comment]
Although I have found AU Lab to be buggy. It’s worth trying though.

**Comment by leocoracini** [comment]
Yeah, AU Lab is kinda buggy. Sometimes doesn't work automatically when plug/unplug earphones. It also buggy input when on Skype meeting. Audio is a must have in my daily life, so I prefer some alternative that "just works" all the time.
I've tried the SoundSource in trial mode, really like it but my currency is Brazilian Real, so the price in US dollars is quite expensive 🙁

Thanks guys!

**Comment by v10x** [comment]
The main issue with AU Lab is that apart from bugs and being generally unreliable, it's required to have its GUI always running and cannot be used in any other way.

What I'd like to see is a way to have AU plugins directly into BlackHole's core, this not only would reduce latency to a minimum, but it'd also be non-intrusive.

**Comment by odinian** [comment]
+1 for AU plugins.

**Comment by bettertallpaul** [comment]
+1 for AU plugins.

**Comment by kkanungo17** [comment]
Yep, addition of AU plugins support would make this the perfect open-source solution for system-wide DSP effects.

**Comment by sound-eng** [comment]
+1 for AU! it's not just that AU LAB is buggy but it's also CPU intensive.

**Comment by fjolnir** [comment]
+1 🙏

**Comment by tom-dreamviewer** [comment]
+1

**Comment by rowe-morehouse** [comment]
Great app, thank you. +1 for AU support. Also, you guys got hunted on Product Hunt. You should go over there and claim it as a maker. https://www.producthunt.com/posts/blackhole-1

**Comment by ProfFan** [comment]
+1, I have AU Labs running quite smoothly but it does sometimes have artifacts and loss of audio output

**Comment by jslpc** [comment]
Would absolutely love AU support

**Comment by freakm941** [comment]
i want this feature!

**Comment by odinian** [comment]
I think Blackhole is very good, but if you want AU integration and are willing to go with commercial software, take a look at SoundSource by Rogue Amoeba

**Comment by rowe-morehouse** [comment]
@odinian … Blackhole and SoundSource are complimentary. You can't replace one with the other — they are not the same.

Blackhole is an audio router (like an advanced Soundflower) while SoundSource allows you to control the volume on individual applications (like a mixer for your apps) and choose different outputs for each app.

For example, if I want to route audio from chrome to record into Audactity, I use SoundSource to route output from Chrome to Blackhole, then use Blackhole as the input on Audacity.

Yes, SoundSource supports AU, you can put effects on any app that is playing sound. It is a great app, worth the $39 in my opinion. I think you'll love it too. 

Also, BLACKHOLE RULES. :)

**Comment by L-C-P** [comment]
+1 for AU plugins.
SoundSource is the wrong competitor. But Loopback from the same company is right one for Blackhole. You get an UI and have to pay some money. In addition you can use Audio Hijack to include AU Plugins to your mic.
But I would really prefer an open source solution :-)

**Comment by rowe-morehouse** [comment]
_a little bit opinionated here:_

After thinking about it, **I don't think BlackHole should support AU**. Keep it simple. [ UNIX philosophy](https://www.wikiwand.com/en/Unix_philosophy). 

Put your plugins somewhere else in the chain. 

I think dev time on BlackHole should be spent on making it tighter, less latency (although I have not experienced problems) … just improve the core of it if you are going to work on it. 

The BlackHole integration with SoundSource is awesome. SoundSource was the best $39 bucks I've spent in a long time.

**Reply by designerfuzzi** [reply]
agree because after AU support you have to support Automation and after Automation, Notation and after Notation you support something completely different than BlackHole was meant to be.

**Reply by ekstasis** [reply]
rowe-morehouse:
> Put your plugins somewhere else in the chain.

Makes me wonder if I could find or make a simple background or menu-based AU host, as a stripped-down, UI-free (while running), and open-source replacement for AU Lab.

**Reply by topazcoding** [reply]
+1 for AU support. for now ju-x is ok but misses out some plugins.

**Reply by designerfuzzi** [reply]
-1 again.. because this feature will potentially mess up all installed DAW's and there VST that try to send audio over to other DAWs via BlackHole. There are an increasing amount of Plugins out there that use Singletons to manage internet connections or OpenSoundControl .. So if this feature would ever make it into BlackHole you are forced to make a really properly made AudioUnit verification before startup or Audio source/destination changes and on top you would have to implement also a preferences list that keeps signatures of Audio Units around that must be ignored. 

I ask myself here, why using BlackHole to route Audio if it would be bend to be a tiny DAW sitting always allocated in the memory that stays in concurrence to any audio software that could benefit from BlackHoles main reason to exist. A completely different talk would be to have a derivate of Software that takes advantage of BlackHole and apply Audio processing on Top. Because such a second Software could manage vst verification lists and do other sweet stuff.. but there you have it, its another App. And for testing purpose i would never use an system hooked audio process that can crash the whole system.. So clearly Blackhole should never become some smaller AU Lab.. There are good developers reasons why AU Lab is so picky with plugins..

**Reply by odinian** [reply]
> 
> The BlackHole integration with SoundSource is awesome. SoundSource was the best $39 bucks I've spent in a long time.

Why would one use BlackHole and Sound Source? SoundSource already captures the audio and provides all sorts of options. BTW, I most use Windows now, and boy do I miss SoundSource. _Nothing_ like it on windows. Nothing.

**Comment by a-zhendi** [comment]
-1000! Please don't mess Blackhole with extra features like plugin support. Adding features like this will lead to more complexity, more dependencies, more bugs and less stability. Keep it simple and strong focusing and reliable. Plugin host must be a separate solution.

**Reply by devgizzo** [reply]
Features like an inline plugin host should be 100% isolated from the default audio chain I think, stability of the audio feed is the most important thing to me. I can't live with a single skip or click artifact and need the lowest latency.

**Comment by RJVB** [comment]
I think this could be very useful, and shouldn't introduce any additional latency or other issues as long as you don't actually use any plugins. Maybe the GUI part could be implemented via an external application that acts as a combination of an insert cable plus pedal board. If you want support for chaining multiple AU plugins; I'd be fine with support for a single one personally.

**Comment by LeXuanKhanh** [comment]
+1 for AU plugins, this is a last feature to make BlackHole a perfect mac audio app

**Comment by devgizzo** [comment]
You can already run system wide vsts using BlackHole. I couldn't get au lab to run on my m1 with latest MacOs. I use BlackHole with LightHost vst host to equalize system audio. https://github.com/rolandoislas/LightHost.

Setup:

System audio out switched to BlackHole, soundcard and blackhole in aggregate, aggregate selected as input on the vst host, soundcard as output on the vst host. In aggregates I found soundcard has to be the clock source and only blackhole should be drift compensated.

Priorities:

As for pops, clicks and drift artifacts I am still testing. I think it is generally okay, but I prefer absolute accurate sound without ANY artifacts or drift above any new features. Is the drift correction not skipping a single bit of audio I wonder?

**Reply by worklitian** [reply]
Thank you for posting your setup. It solved the random pops/drifts on my m1 mac. For reference, I use system audio > BlackHole > Hosting AU app > external DAC, and setting the DAC as clock source worked.

**Reply by devgizzo** [reply]
I still regularly get drift like beat skips either using Blackhole with or without aggregate. It is very irregular. Seems to also depend on sample buffer size, higher number = less skips and/or multi tasking audio between DAW and youtube. I now run one vst in LightHost, which is TDR Nova EQ.

Seems I traded PC DPC Latency clicks for beat skips on MacOS. It is workable, but annoying.
Been thinking about getting a minidsp to get rid of system audio EQ for room correction.

Can anybody tell why it is recommended to create an aggregate with soundcard and blackhole in the first place?
If I set Blackhole as system audio source I can use it as input for LightHost which outputs to my audio interface, all the aggregate gets me is drift correction. Why would audio drift / beat skip when I don't use an aggregate?

**Comment by devinroth** [comment]
That sounds like audio drift. Aggregates allow for drift correction. I don’t have experience with the apps you spoke about but the input and the output will both need to use the aggregate if you want to avoid drift.

**Reply by devgizzo** [reply]
Thank you for responding and all your effort so far on the project, I will fiddle with the setup further and report back my findings.
The VST host I use is this: [https://github.com/rolandoislas/LightHost/releases](https://github.com/rolandoislas/LightHost/releases)
Its written for an old version of MacOS and not supported anymore, so it could also be the problem in this setup ofcourse.

I just switched back to the aggregate I already tried using for system output and use it as input in LightHost. The output of LightHost goes straight to the audio interface. Played 10 minutes of music using Foobar and it was fine. Switched to youtube, it was distorted with flutter. So deleted the aggregate and set up a new one. Now it has been skip free for a while.
It can work flawless for an hour maybe. It's really hard to pinpoint the problem.

First thing I did after installing was create multi output and aggregate, because the manual said so.
But as I found them to be skipping I thought I could skip on the aggregates as I read that audio drift is about correcting the sample rate differences between grouped devices. In this setup I don't have a need to group devices. Unless it is mandatory to prevent drifting!

I did try the aggregate before but it was also skipping at some point. Right now it is flawless.

**Reply by devgizzo** [reply]
Okay, after 2 weeks of testing with both aggregate in and out.
Blackhole selected in the aggregate.
Interface selected in the aggregate.
Clock source to blackhole. 48khz same for vst host.
Drift correction on Blackhole 16ch.
Only one app playing sound. Foobar music player playing a set of m4a files.
No mouse movement. No other app / cpu usage.

With this setup I get one beat skip / audio drift every 7 or 7.5 minutes.
Either with the vst host sending straight to the interface or sending to a multi output with drift control on the interface.
M1 macbook max on latest macos.

**Comment by devinroth** [comment]
Is it an issue with your vst host?

BlackHole is used in a number of applications using drift correction (my own code in some cases) without issue.

Do a test with audacity to see if you can reproduce the problem.

**Comment by devgizzo** [comment]
Maybe this is why I didn't notice any drift while using Bitwig. It is indeed drift free in audacity through the aggregate!

So now I am looking for a new system wide vst host to run my speaker equalization on, any suggestions?
I guess I also upvote the support for a lightweight vst host within blackhole now. :-) Although I would still make it optional and loosely coupled.

**Comment by devinroth** [comment]
Something like that is on my to do list. It wouldn’t be part of BlackHole but rather a separate thing that uses BlackHole.

**Reply by Kentsoon** [reply]
+1 for AU Support.
I use "Sound correction" plugins like Sonarworks SoundID and IK Multimedia ARC3 to calibrate my speakers and headphones to deliver an accurate studio reference sound in my working environment. I use these AU Plugins Systemwide with an App called "menubus" (free version). I do not like using these Plugins in my DAW, because one tends to forget to deactivate them, when printing Audio for Distribution or Mastering, which makes the result unusable.
Unfortunately the App "menubus" is no longer for sale and is not developed/supported any more.
Soundsource or AudioHijack are way too much App for that purpose. And in Soundsource you can not store different AU setups etc. 
The reason why I'm voting for Blackhole:
Zero Latency - Small App.
If you are planing on developing something like that, have a look at menubus.
The Pro version can still be downloaded and is exactly what most people are looking for, when wanting to use AU Plugins sytemwide.

**Comment by vcoras** [comment]
There is another simple way I use to run system-wide audio via an EQ with BlackHole. Currently, I am using [Element](https://github.com/kushview/element) as a VST host. You can build it or download for free.

The setup (tested on Intel and M1 CPUs):

- select BlackHole as Audio Input Device in Element and in macOS as sound source
- select any output device (an aggregate device in my case for MacBook speakers and external DAC)
- make the connections in Element's graph so that the audio chain looks something like: BlackHole ⇾ <VST/VST3/AU plugin> ⇾ Your output device

Example:
<img width="918" alt="image" src="https://github.com/ExistentialAudio/BlackHole/assets/57938881/93fb462a-015d-4a16-a50a-3cfb328caaff">

Previously, before Sonoma, I used to experience random audio dropouts now and then as well as audio drift when resuming from sleep. But now, this setup seems to be pretty stable in macOS Sonoma on MacBook Pro M1 with no pops, audio drift or dropouts.

In conclusion, a plugin host for BlackHole would be a nice feature to have if it wouldn't be for the downsides (complexity, latency, etc.), but one can relatively simply and quickly recreate this setup to achieve the same goal.

**Reply by LeXuanKhanh** [reply]
thanks for mention Element, i'm current using audio hijack for using vst/ au for my mic. This seem like an free alternative for that due to easy-to-use graph interface. Many vst host/DAW i find is kinda out-of-date or very robust to use

**Reply by devgizzo** [reply]
Good to hear it doesn't drift on Sonoma. I have tried a few free vst hosts myself but I got drift and dropouts sometimes only just after hours of stable use. So I ended up ditching aggregates for every day use for now. Still blackhole is great in that you can route 4 separate stereo channels of audio from traktor to bitwig and run your mix decks through bitwig beat synced fx on a single laptop. If you don't open anything else it can work flawless for more than an hour. Still, macos audio drift on aggregates is what DPC latency is on pc. I hope it gets solved for good some day.

**Comment by macuserguru** [comment]
I search for a tool to use headphone settings from Oratory1990 or create, edit EQ settings with editor
I testet Sonarworks (expansive), SoundSource (it's not exactly was I want, I could retest them), eqMac (for full features a license needed).
Is Sonomona the right for this?
My music players an M1 Mac: Foobar2000-v2.6, Fidelia, Adirvana, Media Center. 

Are [removed] the right thing for me?

**Reply by worklitian** [reply]
Autoeq.app is not an app. The page asks you to find your own app to host the eq profiles. For macos it suggests this: "MacOS AUNBandEq with AU Lab and BlackHole or Soundflower. Alternatively paid options eqMac and SoundSource offer easier setup."

Be aware that there are 3 components in the free option: AUNBandEq, AU lab, and BlackHole. You can replace them with other alternatives, but you always need these 3 parts

**Comment by RJVB** [comment]


**Reply by devinroth** [reply]
This discussion does not belong here. Please discuss in the appropriate repo. I am also removing the link due to no open source license.
